File: frontend/src/lib/listingParties.js
export const MAX_CLIENTS = 4

export const DEFAULT_LISTING_PARTIES = {
  agent: { name: 'Walter', email: '', color: '#e0e7ff' },
  clients: [{ name: 'Client', email: '', color: '#fef3c7' }],
}

/** First token of a full name — used for agent display everywhere. */
export function firstNameOnly(name) {
  const t = String(name || '').trim()
  if (!t) return ''
  return t.split(/\s+/)[0]
}

const CLIENT_COLORS = ['#fef3c7', '#fce7f3', '#d1fae5', '#e0f2fe']

export function normalizeListingParties(raw) {
  if (!raw || typeof raw !== 'object') {
    return structuredClone(DEFAULT_LISTING_PARTIES)
  }
  const agent = raw.agent && typeof raw.agent === 'object' ? raw.agent : {}
  const clientsIn = Array.isArray(raw.clients) ? raw.clients : []
  const clients = []
  for (const item of clientsIn) {
    if (!item || typeof item !== 'object') continue
    const name = String(item.name || '').trim()
    if (!name) continue
    clients.push({
      name,
      email: String(item.email || '').trim(),
      color: String(item.color || '').trim() || CLIENT_COLORS[clients.length % CLIENT_COLORS.length],
    })
    if (clients.length >= MAX_CLIENTS) break
  }
  const agentRaw = String(agent.name || DEFAULT_LISTING_PARTIES.agent.name).trim()
  return {
    agent: {
      name: firstNameOnly(agentRaw) || DEFAULT_LISTING_PARTIES.agent.name,
      email: String(agent.email || '').trim(),
      color: String(agent.color || DEFAULT_LISTING_PARTIES.agent.color).trim() || DEFAULT_LISTING_PARTIES.agent.color,
    },
    clients: clients.length ? clients : structuredClone(DEFAULT_LISTING_PARTIES.clients),
  }
}

/** Four editable rows (blank slots allowed in the form). */
export function clientFormRows(parties) {
  const filled = [...(parties?.clients || [])]
  while (filled.length < MAX_CLIENTS) {
    filled.push({
      name: '',
      email: '',
      color: CLIENT_COLORS[filled.length % CLIENT_COLORS.length],
    })
  }
  return filled.slice(0, MAX_CLIENTS)
}

export function partiesForSave(agent, clientRows) {
  const clients = clientRows
    .map((row) => ({
      name: String(row.name || '').trim(),
      email: String(row.email || '').trim(),
      color: String(row.color || '').trim(),
    }))
    .filter((row) => row.name)
  return {
    agent: {
      name: firstNameOnly(String(agent.name || '').trim()) || DEFAULT_LISTING_PARTIES.agent.name,
      email: String(agent.email || '').trim(),
      color: String(agent.color || '').trim() || DEFAULT_LISTING_PARTIES.agent.color,
    },
    clients: clients.length ? clients : DEFAULT_LISTING_PARTIES.clients,
  }
}

export function agentDisplayName(parties) {
  const raw = parties?.agent?.name?.trim() || DEFAULT_LISTING_PARTIES.agent.name
  return firstNameOnly(raw) || DEFAULT_LISTING_PARTIES.agent.name
}

export function clientDisplayNames(parties) {
  const names = (parties?.clients || []).map((c) => c.name?.trim()).filter(Boolean)
  return names.length ? names : [DEFAULT_LISTING_PARTIES.clients[0].name]
}

export function clientNamesLabel(parties) {
  const names = clientDisplayNames(parties)
  if (names.length === 1) return names[0]
  if (names.length === 2) return `${names[0]} and ${names[1]}`
  return `${names.slice(0, -1).join(', ')}, and ${names[names.length - 1]}`
}

File: backend/app/listing_parties.py
import json
from typing import Any

def _first_name_only(name: str) -> str:
    t = (name or "").strip()
    if not t:
        return ""
    return t.split()[0]


DEFAULT_LISTING_PARTIES: dict[str, Any] = {
    "agent": {"name": "Walter", "email": "", "color": "#e0e7ff"},
    "coordinator": {"name": "", "email": "", "color": "#ddd6fe"},
    "clients": [{"name": "Client", "email": "", "color": "#fef3c7"}],
}

MAX_CLIENTS = 4


def _parse_person(raw: dict | None, defaults: dict) -> dict[str, str]:
    raw = raw if isinstance(raw, dict) else {}
    return {
        "name": str(raw.get("name") or defaults.get("name") or "").strip(),
        "email": str(raw.get("email") or "").strip(),
        "color": str(raw.get("color") or defaults.get("color") or "#fef3c7").strip()
        or defaults.get("color", "#fef3c7"),
    }


def parse_listing_parties(raw: str | None) -> dict[str, Any]:
    if not raw or not str(raw).strip():
        return json.loads(json.dumps(DEFAULT_LISTING_PARTIES))
    try:
        data = json.loads(raw)
    except json.JSONDecodeError:
        return json.loads(json.dumps(DEFAULT_LISTING_PARTIES))
    agent = data.get("agent") if isinstance(data.get("agent"), dict) else {}
    coordinator = data.get("coordinator") if isinstance(data.get("coordinator"), dict) else {}
    clients_in = data.get("clients") if isinstance(data.get("clients"), list) else []
    clients = []
    for item in clients_in[:MAX_CLIENTS]:
        if not isinstance(item, dict):
            continue
        name = str(item.get("name") or "").strip()
        if not name:
            continue
        clients.append(
            {
                "name": name,
                "email": str(item.get("email") or "").strip(),
                "color": str(item.get("color") or "#fef3c7").strip() or "#fef3c7",
            }
        )
    agent_raw = str(agent.get("name") or DEFAULT_LISTING_PARTIES["agent"]["name"]).strip()
    coord = _parse_person(coordinator, DEFAULT_LISTING_PARTIES["coordinator"])
    return {
        "agent": {
            "name": _first_name_only(agent_raw) or DEFAULT_LISTING_PARTIES["agent"]["name"],
            "email": str(agent.get("email") or "").strip(),
            "color": str(agent.get("color") or DEFAULT_LISTING_PARTIES["agent"]["color"]).strip()
            or DEFAULT_LISTING_PARTIES["agent"]["color"],
        },
        "coordinator": coord,
        "clients": clients or list(DEFAULT_LISTING_PARTIES["clients"]),
    }


def dump_listing_parties(data: dict[str, Any]) -> str:
    parsed = parse_listing_parties(json.dumps(data))
    return json.dumps(parsed)
